### PR TITLE
student page filtering

### PR DESCRIPTION
**Pull request by emmaeagles**
# Summary

Added filters to the view students page

## Test Plan

How did you test your changes?

it works :-)
@paul needs to make the controller class for studetn before i can make tests fir the controller methods but i did do service tests

## Related Issues

Which issues does this PR resolve/work on?
#135

### CODE REVIEW

File: Backend/src/main/java/ca/mcgill/cooperator/controller/ControllerUtils.java [hunk 1]
```diff
@@ -560,6 +560,20 @@ public static CourseOfferingDto convertToDto(CourseOffering co) {
         }
         return courseOfferingDtos;
     }
+    
+    public static Set<CourseOfferingDto> convertCourseOfferingSetToDto(
+            Set<CourseOffering> courseOfferings) {
```
Comment (kraglalbert) [inline Backend/src/main/java/ca/mcgill/cooperator/controller/ControllerUtils.java]: Change `Set` to `Collection` so that it's more general and works for both sets and lists

File: Backend/src/main/java/ca/mcgill/cooperator/controller/ControllerUtils.java [hunk 2]
```diff
@@ -1011,7 +1025,19 @@ public static StudentDto convertToDto(Student s) {
         return studentDto;
     }
 
-    public static List<StudentDto> convertToDto(List<Student> students) {
+    public static Set<StudentDto> convertStudentSetToDto(Set<Student> students) {
+        Set<StudentDto> studentDtos = new HashSet<StudentDto>();
```
Comment (kraglalbert) [inline Backend/src/main/java/ca/mcgill/cooperator/controller/ControllerUtils.java]: Same thing here

File: Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java [hunk 3]
```diff
@@ -38,10 +43,10 @@ public CourseOfferingDto getCourseOfferingById(@PathVariable int id) {
     }
 
     @GetMapping("/course/{id}")
-    public List<CourseOfferingDto> getCourseOfferingsByCourse(@PathVariable int id) {
+    public Set<CourseOfferingDto> getCourseOfferingsByCourse(@PathVariable int id) {
```
Comment (kraglalbert) [inline Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java]: Why did you change this to a set?

File: Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java [hunk 4]
```diff
@@ -61,6 +66,22 @@ public CourseOfferingDto createCourseOfferingDto(
...
     }
+    
+    @GetMapping("/seasons")
+    public Season[] getAllCoopSeasons() {
+        Season[] seasons = Season.values();
+        return seasons;
+    }
+    
+    @GetMapping("/years")
+    public int[] getYearsToDisplay() {
```
Comment (kraglalbert) [inline Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java]: Making a backend call for this is probably overkill, we can just make a function for this in the frontend (in `boot/common.js`)

File: Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java [hunk 5]
```diff
@@ -40,7 +48,48 @@
...
+     * @return List<StudentDto>
+     */
+    @GetMapping("/filter")
+    public Set<StudentDto> getStudentFiltered(
+    		@RequestParam Season season, 
+    		@RequestParam Integer year, 
+    		@RequestParam String name, 
+    		@RequestParam CoopStatus status) {
+    	Set<CourseOffering> result = courseOfferingService.getAllCourseOfferingsSet();
+    	if(year != null) {
```
Comment (kraglalbert) [inline Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: You can combine this with the method above and make the params optional; if they're not present then return all students, otherwise do the filtering (this is better design from an external API perspective so that we have less routes)

File: Frontend/src/pages/admin/AdminStudents.vue [hunk 6]
```diff
@@ -1,8 +1,38 @@
 <template>
```
Comment (kraglalbert) [inline Frontend/src/pages/admin/AdminStudents.vue]: nit: change the file name to `AdminStudentsPage.vue` so it's more obvious that it's a page when you have it open in an editor. We didn't do this at first but I've started doing it because I think it's more clear

File: Frontend/src/pages/admin/AdminStudents.vue [hunk 7]
```diff
@@ -83,10 +110,81 @@ export default {
       .then(resp => {
         this.students = resp.data;
       });
+    this.$axios
+      .get("/coops/status", {
+        headers: {
+          Authorization: this.$store.state.token
+        }
```
Comment (kraglalbert) [inline Frontend/src/pages/admin/AdminStudents.vue]: Don't need this yet technically since we haven't implemented actual authentication

File: Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java (hunk 3 above)
Comment (emmaeagles) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java]: because i need to do the set intersections and you cant use lists for that

Comment (kraglalbert) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/CourseOfferingController.java]: Ah right okay makes sense

File: Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java (hunk 5 above)
Comment (emmaeagles) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: i dont think this is that good of an idea now that im trying to implement bc it means i have to check to make sure that all the filter inputs were null/not there to return students which is a whole other portion of the method with just a huge if statement 

i thought at first it would just be if the size of the result is 0 then just return all but thats wrong bc the combo of filters could just return no matching records soooo i can combine them but its not as slick as i originally thought

Comment (emmaeagles) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: image

IDC i can do this if you want

**Comment by emmaeagles** [comment]
image

No filter

File: Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java (hunk 5 above)
Comment (emmaeagles) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: ok im kidding it doesnt work when its this way

File: Frontend/src/pages/admin/AdminStudents.vue (hunk 7 above)
Comment (emmaeagles) [reply Frontend/src/pages/admin/AdminStudents.vue]: so i should remove them all? aren't we going to need this eventually

Comment (kraglalbert) [reply Frontend/src/pages/admin/AdminStudents.vue]: You can leave it but just FYI for future stuff that you don't need to add this

File: Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java (hunk 5 above)
Comment (kraglalbert) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: Why doesn't it work? The code looks ok as far as I can tell

Comment (emmaeagles) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: Lol it turns out there was a bug in the filter and it was EXPOSED when i added that line but all good i fixed it

Comment (kraglalbert) [reply Backend/src/main/java/ca/mcgill/cooperator/controller/StudentController.java]: Ok nice I would say keep it this way as one route then

**Comment by emmaeagles** [comment]
image

Filter on status

**Comment by emmaeagles** [comment]
image

filter on all categories

File: Frontend/src/boot/common.js [hunk 8]
```diff
@@ -3,7 +3,16 @@ import Vue from "vue";
 /*
  * This object contains functions that are globally available to every component
  */
+
 const common = {
+  getYears: function() {
+    var current = new Date().getFullYear();
+    var yrs = [];
+    for (var i = 0; i < 10; i++) {
+      yrs.push(current - i);
```
Comment (kraglalbert) [inline Frontend/src/boot/common.js]: I think this should also have at least 1 year ahead of the current year

File: Frontend/src/pages/admin/AdminStudentsPage.vue [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        :columns="columns"
+        row-key="studentName"
+        @row-click="goToStudentCoop"
+      />
+    </div>
+  </div>
+</template>
+<script>
+export default {
+  name: "AdminAllStudents",
```
Comment (kraglalbert) [inline Frontend/src/pages/admin/AdminStudentsPage.vue]: This should match the file name
